### PR TITLE
Add some documentation and tests for goplugins

### PR DESCRIPTION
**Pull request by alephnull**
Uses docker-compose

## Description
A Makefile and some documentation to test goplugins. This is for gateway versions _after_ 2.9.4.1 as prior versions exposed `plugin-build` and will probably not work.

## Related Issue
#3195 

## Motivation and Context
Testing goplugins is cumbersome as it requires a bit of setup and the commands are long and easy to mess up.

## How This Has Been Tested
By hand.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Refactoring or add test (improvements in base code or adds test coverage to functionality)

## Checklist

- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). If pulling from your own
      fork, don't request your `master`!
- [x] Make sure you are making a pull request against the **`master` branch** (left side). Also, you should start
      *your branch* off *our latest `master`*.
- [ ] My change requires a change to the documentation.
  - [ ] If you've changed APIs, describe what needs to be updated in the documentation.
  - [ ] If new config option added, ensure that it can be set via ENV variable
- [x] I have updated the documentation accordingly.
- [ ] Modules and vendor dependencies have been updated; run `go mod tidy && go mod vendor`
- [ ] When updating library version must provide reason/explanation for this update.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] Check your code additions will not fail linting checks:
  - [ ] `go fmt -s`
  - [ ] `go vet`

### CODE REVIEW

File: images/plugin-compiler/README.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+```shellsession
+% docker run -v `pwd`:/plugin-source tykio/tyk-plugin-compiler:v2.9.4.2 myplugin.so
+```
+
+You will find a myplugin.so in the current directory which is the file
+that goes into the API definition
+
+## Plugin aliasing
+Due to the way that `dl_open(3)` works, the filename needs to change
```
Comment (letzya) [inline images/plugin-compiler/README.md:22]: Are you referring to the fact you need to give the name `myplugin.so` when compiling? It's not so clear

File: images/plugin-compiler/README.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+mount your code into the image. Since the vendor directory needs to be
+identical between the gateway build and the plugin build, this means
+that you should pull the version of this image corresponding to the
+gateway version you are using.
+
+This also implies that if your plugin has vendored modules that are
+also used by Tyk gateway then your module will be overridden by the
+version that Tyk uses.
+
+```shellsession
```
[the remark this answers is not in the document]
Comment (alephnull) [reply images/plugin-compiler/README.md:14]: ?

File: images/plugin-compiler/README.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+## Plugin aliasing
+Plugins are loaded via `dl_open(3)` and the shared library cache,
+`ld.so.cache` will be used. Therefore,even if a plugin's content
+changes but the filename does not, the cached plugin will be used.
+
+See the manpages for `dl_open(3)` and `ld.so(8)` on your platform for
+more details.
+
+## Hot reloading
```
Comment (letzya) [inline images/plugin-compiler/README.md:29]: Maybe I missed it but you explain how not to reload a plugin. Can we explain what we should do when we want to update the plugin?
(I assume- same name and restart or new name and hot reload?)

File: images/plugin-compiler/README.md [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+Using `/tyk/reload/group` will _not_ update the plugin if you have
+compiled a fresh version and used the same pathname.
+
+To reload a plugin, you will have to restart the gateway process.
```
Comment (letzya) [inline images/plugin-compiler/README.md:33]: This could open a sea of complaints if that’s the only way @alephnull

Comment (matiasinsaurralde) [reply images/plugin-compiler/README.md:33]: This is hard to fix unless we implement our own dynamic loading logic. We should wait for the `plugin` package to implement something that allows the safe unloading/reloading of Go shared libraries, etc.
For now the only strategy is to use different filenames every time the code changes or perform a full process reload.

File: images/plugin-compiler/README.md [hunk 5]
```diff
@@ -0,0 +1,33 @@
+# Tyk Plugin compiler
+
+## Building a plugin
+Navigate to where your plugin is and build using a docker volume to
```
Comment (matiasinsaurralde) [inline images/plugin-compiler/README.md:4]: Docker

**Pull request closed without merging**